pss/message/message_test.go

```diff
@@ -0,0 +1,51 @@
+package message_test
+
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/epiclabs-io/ut"
```

Comment:
A question: Is there some risk associated with using the same code that is to be tested as the source of truth for the results it is supposed to give?

I am aware that the way we are doing it currently, eg. by copying the new hash from terminal output and pasting it into the test, probably doesn't provide the best guarantee either. I just can't help wondering that automation somehow invites the dingo.

Reply:
That aside:

If I remember correctly, our conclusion from discussing this half a year ago was that @jpeletier should feel free to try out the tool confined to the feeds package, as this more or less was a package he alone was developing anyway.

Also, I think the controversy concerning the package was _not_ the test result data generation part of it, but rather that the assertion tools seemed superfluous. IIRC @nonsense was particularly outspoken there.

I am not against trying this out in pss instead, as @kortatu and @aperonam and he all are familiar with the framework and actively make use of it in their development work in general. But perhaps we for now should confine it to this package, at least before we have another team-wide discussion about introducing it.

Reply:
> A question: Is there some risk associated with using the same code that is to be tested as the source of truth for the results it is supposed to give?

It is not exactly like this. The way to use it is: you run the test code for the first time, it produces a set of results **which is your responsibility to manually check** and sign off to make sure they make sense (just as you now would type the expected string or value, or calculate the expected hash somewhere else, which you then copy and paste in your code). Once you're happy with the values, you then "freeze" that behavior.

> [...] probably doesn't provide the best guarantee either. I just can't help wondering that automation somehow invites the dingo.

You are not forced to have automated generation, you can type up the results file by hand.

In my opinion the net result is: less time actually writing test code, more time to write for more coverage and test quality.

Reply:
> [...] the assertion tools seemed superfluous. [...]
That is a point of view. My point of view is that any overhead work involved in setting the stage for a condition to be tested a) Takes time to write b) is prone to errors in the test code itself, which takes also time to debug and c) Hides the purpose of the test in the "bureaucracy". All of these make the test harder to read and evolve over time.

> If I remember correctly, our conclusion from discussing this half a year ago was that @jpeletier should feel free to try out the tool confined to the feeds package, as this more or less was a package he alone was developing anyway.

Yes, I am moving that freedom to pss instead, this is not intended to extend generally outside of this package until discussed in depth.

Reply:
guys, I strongly resist this. My 2cents:

- it is generally against go philosophy - readability and focussed tests can be achieved by setup helpers etc,
- such (esp branded) dependencies are unnecessary risk factor due to maintanance/updates/continuity/compatibility
- any mediation in test code is extra risk, because errors can be deeply hidden in the framework (your point b) --- so testing frameworks should only be used in special cases where they genuinely bring needed functionality (network simulation)
- might have saved your team time but would create extra work for the current team  -- so time to learn/master (vs  your point a)
- i would like to see convergence of code/test style not more divergence, so i am not very keen  on introducing it in one place
- autogenerated or not, having few byte files to test a simple serialisation is no better than inline 
-  changing asset directory names if you change testname also not too good

I'd like to hear others opinions, @janos , @lmars @nonsense , so i add it to the roundtable topics for tuesday. https://hackmd.io/oj9_cT2KQimMdIPe_W_ejQ?both

Reply:
One of the main reasons I use Go is because of the simplicity and heavy focus in the community of doing things one way with little focus on customisation. For example, whilst I prefer double spacing to tabs, I use tabs, because that's what `gofmt` does, and everyone uses (or should use :wink:) `gofmt`.

Everyone doing things one way means I can jump into any codebase and immediately feel familiar with the coding style and tooling used and focus on the business logic. This is an important property for any sizeable software project to have where developer onboarding / churn becomes more of a concern over time and focus on business logic rather than bike-shedding is a must.

Applying this to the idea of "test helper" packages, I typically avoid them because the stdlib `testing` package is what most people use, and it gives me what I need to write effective tests.

From a cursory glance at the microTest README, I see things like:

```go
Ok(t, err)

Assert(t, x == y, "expected x to equal %v, got %v", y, x)

Equals(t, x, y)
```

which I would immediately argue should just be:

```go
if err != nil {
	t.Fatal(err)
}

if x != y {
	t.Fatalf("expected x to equal %v, got %v", y, x)
}

if x != y {
	t.Fatal("values aren't equal")
}
```

They may require a few more lines and, for those without editor macros, a few more key strokes, but I'd argue the time / space lost is far outweighed by the clarity of doing it the same way everyone else does it.

On this point:

> one important feature is the ability to generate and store test results for later verification

I personally dislike auto-fixture generation style tests, again because I feel it is simpler and more readable to just have a function at the start of a test that either declares or generates the fixture data. I see no issue with this, for example:

```go
func TestFoo(t *testing.T) {
	var expectedFoo = []byte(`<some fixture data>`)
	foo, err := Foo()
	if err != nil {
		t.Fatal(err)
	}
	if foo != expectedFoo {
		t.Fatalf("expected foo to be %v, got %v", expectedFoo, foo)
	}
}
```

This is my 2cents from just reading the above comment thread.

Reply:
This is only my point of view about testing in Go. And I saw that the third party package was removed.

Using testing package should be the preferred way instead a generalized helper functions in a third party utility package. Simple checks like Ok, Assert, Equal do not bring so much clarity, as it adds an indirection. Helper functions should be written for the package that is tested for abstracting more complex validations. If the package need to provide some unified way of testing other implementations or its types or behaviour in other packages, then a dedicated testing package for this purpose should be made. The generalized testing package on the project level should be as small as possible without dependencies on packages in that project.

Reply:
I have removed these dependencies. If there are no other outstanding concerns, this can be closed.